File: nodes/merge-messages/merge-messages.process.py
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        messages = inputs.get('messages')
        if not isinstance(messages, list):
            raise Exception("Messages input must be an array")

        separator = settings.get('separator', ' ')
        preserve_structure = settings.get('preserve_structure', True)

        if len(messages) == 0:
            return {
                "messages": [],
                "original_count": 0,
                "merged_count": 0,
                "merges_performed": 0
            }

        # Helper function to merge content
        def merge_content(content1, content2):
            if isinstance(content1, str) and isinstance(content2, str):
                return content1 + separator + content2
            elif isinstance(content1, list) and isinstance(content2, list):
                if preserve_structure:
                    # Preserve multi-modal structure
                    return content1 + content2
                else:
                    # Flatten to text
                    text1 = ''.join([item['text'] for item in content1 if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item])
                    text2 = ''.join([item['text'] for item in content2 if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item])
                    return text1 + separator + text2
            elif isinstance(content1, str) and isinstance(content2, list):
                if preserve_structure:
                    return [{'type': 'text', 'text': content1}] + content2
                else:
                    text2 = ''.join([item['text'] for item in content2 if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item])
                    return content1 + separator + text2
            elif isinstance(content1, list) and isinstance(content2, str):
                if preserve_structure:
                    return content1 + [{'type': 'text', 'text': content2}]
                else:
                    text1 = ''.join([item['text'] for item in content1 if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item])
                    return text1 + separator + content2
            else:
                # Fallback - convert to strings
                return str(content1) + separator + str(content2)

        result = []
        merges_performed = 0
        current_message = None

        for message in messages:
            if not isinstance(message, dict) or 'role' not in message:
                # Invalid message - add as-is
                if current_message:
                    result.append(current_message)
                    current_message = None
                result.append(message)
                continue

            if current_message is None:
                # Start new message
                current_message = message.copy()
            elif current_message['role'] == message['role']:
                # Same role - merge content
                current_message['content'] = merge_content(current_message['content'], message['content'])
                merges_performed += 1
            else:
                # Different role - finish current message and start new one
                result.append(current_message)
                current_message = message.copy()

        # Add the last message if there is one
        if current_message:
            result.append(current_message)

        return {
            "messages": result,
            "original_count": len(messages),
            "merged_count": len(result),
            "merges_performed": merges_performed
        }

    except Exception as error:
        print(f'Merge Messages error: {error}')
        raise Exception(f"Merge Messages error: {str(error)}")

File: sdks/nodejs/nodes/merge-messages/merge-messages.process.js
export default async ({inputs, settings, config}) => {
    try {
        const messages = inputs.messages;
        if (!Array.isArray(messages)) {
            throw new Error("Messages input must be an array");
        }

        const separator = settings.separator || " ";
        const preserveStructure = settings.preserve_structure !== false;

        if (messages.length === 0) {
            return {
                messages: [],
                original_count: 0,
                merged_count: 0,
                merges_performed: 0
            };
        }

        // Helper function to merge content
        const mergeContent = (content1, content2) => {
            if (typeof content1 === 'string' && typeof content2 === 'string') {
                return content1 + separator + content2;
            } else if (Array.isArray(content1) && Array.isArray(content2)) {
                if (preserveStructure) {
                    // Preserve multi-modal structure
                    return [...content1, ...content2];
                } else {
                    // Flatten to text
                    const text1 = content1.filter(item => item && item.type === 'text' && item.text).map(item => item.text).join('');
                    const text2 = content2.filter(item => item && item.type === 'text' && item.text).map(item => item.text).join('');
                    return text1 + separator + text2;
                }
            } else if (typeof content1 === 'string' && Array.isArray(content2)) {
                if (preserveStructure) {
                    return [{type: 'text', text: content1}, ...content2];
                } else {
                    const text2 = content2.filter(item => item && item.type === 'text' && item.text).map(item => item.text).join('');
                    return content1 + separator + text2;
                }
            } else if (Array.isArray(content1) && typeof content2 === 'string') {
                if (preserveStructure) {
                    return [...content1, {type: 'text', text: content2}];
                } else {
                    const text1 = content1.filter(item => item && item.type === 'text' && item.text).map(item => item.text).join('');
                    return text1 + separator + content2;
                }
            } else {
                // Fallback - convert to strings
                return String(content1) + separator + String(content2);
            }
        };

        const result = [];
        let mergesPerformed = 0;
        let currentMessage = null;

        for (const message of messages) {
            if (!message || typeof message !== 'object' || !message.role) {
                // Invalid message - add as-is
                if (currentMessage) {
                    result.push(currentMessage);
                    currentMessage = null;
                }
                result.push(message);
                continue;
            }

            if (!currentMessage) {
                // Start new message
                currentMessage = { ...message };
            } else if (currentMessage.role === message.role) {
                // Same role - merge content
                currentMessage.content = mergeContent(currentMessage.content, message.content);
                mergesPerformed++;
            } else {
                // Different role - finish current message and start new one
                result.push(currentMessage);
                currentMessage = { ...message };
            }
        }

        // Add the last message if there is one
        if (currentMessage) {
            result.push(currentMessage);
        }

        return {
            messages: result,
            original_count: messages.length,
            merged_count: result.length,
            merges_performed: mergesPerformed
        };

    } catch (error) {
        console.log('Merge Messages error:', error);
        throw new Error(`Merge Messages error: ${error.message}`);
    }
};
